modules/reactor/src/meshgenerators/ControlDrumMeshGenerator.C

```diff
@@ -0,0 +1,517 @@
+  std::vector<Real> drum_radii = std::vector<Real>({_drum_inner_radius, _drum_outer_radius});
+  declareMeshProperty(RGMB::drum_radii, drum_radii);
+  declareMeshProperty(RGMB::is_homogenized, false);
+  declareMeshProperty(RGMB::is_single_pin, false);
+}
+
+std::unique_ptr<MeshBase>
+ControlDrumMeshGenerator::generate()
+{
+  // Must be called to free the ReactorMeshParams mesh
+  freeReactorMeshParams();
+
+  // If bypass_mesh is true, return a null mesh. In this mode, an output mesh is not
+  // generated and only metadata is defined on the generator, so logic related to
+  // generation of output mesh will not be called
+  if (getReactorParam<bool>(RGMB::bypass_meshgen))
+  {
+    auto null_mesh = nullptr;
+    return null_mesh;
+  }
+
+  // This generate() method will be called once the subgenerators that we depend on are
+  // called. This is where we reassign subdomain ids/name in case they were merged when
+  // stitching pins into an assembly. This is also where we set region_id and
+  // assembly_type_id element integers.
+
+  // Define all extra element names and integers
+  std::string plane_id_name = "plane_id";
+  std::string region_id_name = "region_id";
+  std::string pin_type_id_name = "pin_type_id";
+  std::string assembly_type_id_name = "assembly_type_id";
+  const std::string default_block_name =
+      RGMB::DRUM_BLOCK_NAME_PREFIX + std::to_string(_assembly_type);
+
+  auto pin_type_id_int = getElemIntegerFromMesh(*(*_build_mesh), pin_type_id_name);
+  auto region_id_int = getElemIntegerFromMesh(*(*_build_mesh), region_id_name);
+  auto assembly_type_id_int = getElemIntegerFromMesh(*(*_build_mesh), assembly_type_id_name);
+
+  unsigned int plane_id_int = 0;
+  if (_extrude)
+    plane_id_int = getElemIntegerFromMesh(*(*_build_mesh), plane_id_name, true);
+
+  // Get next free block ID in mesh in case subdomain ids need to be remapped
+  auto next_block_id = MooseMeshUtils::getNextFreeSubdomainID(*(*(_build_mesh)));
+  std::map<std::string, SubdomainID> rgmb_name_id_map;
+
+  // Loop through all mesh elements and set region ids and reassign block IDs/names
+  // if they were merged during pin stitching
+  for (auto & elem : (*_build_mesh)->active_element_ptr_range())
+  {
+    elem->set_extra_integer(assembly_type_id_int, _assembly_type);
+    const dof_id_type z_id = _extrude ? elem->get_extra_integer(plane_id_int) : 0;
+
+    // Assembly peripheral element (background / duct), set subdomains according
+    // to user preferences and set pin type id to RGMB::MAX_PIN_TYPE_ID - peripheral index
+    // Region id is inferred from z_id and peripheral_idx
+    const auto base_block_id = elem->subdomain_id();
+    const auto base_block_name = (*_build_mesh)->subdomain_name(base_block_id);
+
+    // Check if block name has correct prefix
+    std::string prefix = RGMB::DRUM_BLOCK_NAME_PREFIX + std::to_string(_assembly_type) + "_R";
+    if (!(base_block_name.find(prefix, 0) == 0))
+      continue;
```

Comment:
when does this happen? should it even happen? Seems we are adding the prefix everywhere?

Reply:
This is added in case there's an element that was not created by RGMB. I haven't worked with this specific case but this line makes sure only modifies elements that have the prefix (created by RGMB)